draft-ietf-tls-esni.md

```diff
@@ -446,33 +491,120 @@ to harmonize these to make sure that we maintain key separation.]]
 
 This value is placed in an "encrypted_server_name" extension.
 
-The client MAY either omit the "server_name" extension or provide
-an innocuous dummy one (this is required for technical conformance
-with {{!RFC7540}}; Section 9.2.)
-
-If the server does not negotiate TLS 1.3 or above, then the client
-MUST abort the connection with an "unsupported_version" alert.
-If the server does not provide an "encrypted_server_name" extension
-in EncryptedExtensions, the client MUST abort the connection with
-an "illegal_parameter" alert. Moreover, it MUST check that
-ClientESNIInner.nonce matches the value of the
-"encrypted_server_name" extension provided by the server,
-and otherwise abort the connection with an "illegal_parameter"
-alert.
-
-## Client-Facing Server Behavior
+The client MUST place the value of ESNIKeys.public_name in the "server_name"
+extension. (This is required for technical conformance with {{!RFC7540}};
+Section 9.2.)
+
+### Handling the server response {#handle-server-response}
+
+If the server negotiates TLS 1.3 or above and provides an
+"encrypted_server_name" extension in EncryptedExtensions, the client
+then processes the extension's "response_type" field:
+
+- If the value is "esni_accept", the client MUST check that the extension's
+  "nonce" field matches ClientESNIInner.nonce and otherwise abort the
+  connection with an "illegal_parameter" alert. The client then proceeds
+  with the connection as usual, verifying the certificate against the desired
+  name.
+
+- If the value is "esni_retry_request", the client proceeds with the handshake,
+  verifying the certificate against ESNIKeys.public_name as described in
+  {{verify-public-name}}. If verification or the handshake fails, the client
+  MUST return a failure to calling application. It MUST NOT use the retry keys
+  as described below.
+
+  Otherwise, when the handshake completes successfully with the public name
+  verified, the client MUST abort the connection with an "esni_required" alert.
+  The client can then regard the ESNI keys as securely replaced by the server.
+  It processes the "retry_keys" field from the server's "encrypted_server_name"
+  extension. If one of the values used a version known to the client, the client
+  SHOULD retry the handshake with a new transport connection, using that value
+  to encrypt the SNI. If no value is applicable, the client SHOULD retry with
+  ESNI disabled.
+
+  These retry keys may only be applied to the retry connection. The client MUST
+  continue to use the previously cached keys for subsequent connections. This
```

Comment:
Does "previously cached keys" refer to the keys fetched from DNS?

Reply:
@martinthomson left a bunch of comments that suggested he didn't want the specification text to talk about DNS keys. Thus, to answer your question, yes but not necessarily so we can't say that. :-P

On the list, someone thought a version of this text implied that you're supposed to keep track of all the retry keys and make sure they never repeat. (Honestly, I think that reading was tenuous to say the least.) So I've tried to avoid it saying MUST NOT use the old keys.

I've now switched it to "previously-advertised" which matches the verb used elsewhere. If you all still don't like it, please give me an alternative! :-)